src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -2045,13 +2105,26 @@ private void ReturnConnectionToPool()
             {
                 Debug.Assert(!_detachedFromPool, "Should not be detached from pool unless _connectionClose is true");
 
-                _idleSinceTickCount = Environment.TickCount64;
-
                 // Put connection back in the pool.
                 _pool.RecycleHttp11Connection(this);
             }
         }
 
+        // This should be called while holding the lock (mustn't overlap with SendAsync).
+        internal void OnAddingIdleConnectionToPool()
+        {
+            _idleSinceTickCount = Environment.TickCount64;
+
+            // It could be a while until we start serving the next request. Issue a read to
+            // the transport stream so we can quickly react to potential server disconnects.
+            if (_lastRequestWasAsync)
+            {
+                Debug.Assert(!ReadAheadTaskHasStarted());
+
+                EnsureReadAheadTaskHasStarted();
+            }
```

Comment:
For cases where the connection will be reused very soon, doesn't this add at least some additional overhead?  Seems like we'll end up issuing a zero-byte read we wouldn't otherwise issue, and incur an additional async state machine allocation, yes?

What does doing this aggressively significantly improve?  Have we received complaints about it?

Reply:
Yes, it would mean an additional allocation for the state machine if the connection is reused soon, but not immediately (that is, if we don't have requests already queued on the pool).

The benefit is that we'll immediately get rid of resources if the connection is closed (without waiting for scavenging logic). I can't say whether it's significant (and I don't recall any customer complaints about it) - the main idea behind it is #60729.

> To address https://github.com/dotnet/runtime/issues/60729, we also issue the read-ahead once a connection becomes idle (gets returned to the pool with no remaining requests). If we don't want to do this it's a matter of deleting [this](https://github.com/MihaZupan/runtime/blob/c52af6a0b74ce98f189fe9a8202e14dbedbe0bf2/src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L2111-L2118).

If we want to solve #60729, we pretty much have to have an outstanding read to the transport.
But there is some room regarding how proactive we'd want to be. This PR is about as immediate as we can get, but maybe it would be enough to make scavenging logic more aggressive, if we care.

If we don't care about #60729 (or at least not as strictly as this PR does), we can avoid this allocation.

Reply:
I suggest unbundling it from this change.  We can have a separate discussion about whether there's actually something we need to fix, and if so, how.